cirq/study/flatten_expressions_test.py

```diff
@@ -0,0 +1,182 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sympy
+import cirq
+from cirq.study import flatten_expressions
+
+
+def test_expr_map_names():
+    flattener = flatten_expressions._ParamFlattener({'collision': '<x + 2>'})
+    expressions = [sympy.Symbol('x') + i for i in range(3)]
+    syms = flattener.flatten(expressions)
+    assert syms == [
+        sympy.Symbol(name) for name in ('x', '<x + 1>', '<x + 2>_1')
+    ]
+
+
+def test_flattener_value_of():
+    flattener = flatten_expressions._ParamFlattener({'c': 5, 'x1': 'x1'})
+    assert flattener.value_of(9) == 9
+    assert flattener.value_of('c') == 5
+    assert flattener.value_of(sympy.Symbol('c')) == 5
+    # Twice
+    assert (flattener.value_of(sympy.Symbol('c') / 2 +
+                               1) == sympy.Symbol('<c/2 + 1>'))
+    assert (flattener.value_of(sympy.Symbol('c') / 2 +
+                               1) == sympy.Symbol('<c/2 + 1>'))
+    # Collisions
+    assert (flattener.value_of(
+        sympy.Symbol('c') /
+        sympy.Symbol('2 + 1')) == sympy.Symbol('<c/2 + 1>_1'))
```

Comment:
This is a rather weird use case, since c/(2+1) is different than c/2 +1.  I'm not sure about this behavior.

In any case, this seems like something we would not want to explicitly test.

Reply:
But this test is testing collisions.  It's not clear that we want to enforce via testing that 'c'/'2+1' should collide with 'c'/2+1.  The two seem like different expressions to me.

I would actually qualify this as surprising behavior, since it is really a collision of the symbol of '2+1' with 2+1

Reply:
This is a sympy issue.  Ideally `str(symbol1) == str(symbol2)` implies `symbol1 == symbol2` and it does if your symbol names aren't unusual.  Would you prefer that we add checks that `symbol == sympy.sympify(symbol.name)` to avoid this?